drivers/ethernet/eth_e1000.c

```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright (c) 2018 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define SYS_LOG_DOMAIN "ETH E1000"
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_ETHERNET_LEVEL
+#include <logging/sys_log.h>
+
+#include <zephyr.h>
+#include <net/ethernet.h>
+#include <pci/pci.h>
+#include "eth_e1000_priv.h"
+
+static const char *e1000_reg_to_string(enum e1000_reg_t r)
+{
+#define _(_x)	case _x: return #_x
+	switch (r) {
+	_(CTRL);
+	_(ICR);
+	_(ICS);
+	_(IMS);
+	_(RCTL);
+	_(TCTL);
+	_(RDBAL);
+	_(RDBAH);
+	_(RDLEN);
+	_(RDH);
+	_(RDT);
+	_(TDBAL);
+	_(TDBAH);
+	_(TDLEN);
+	_(TDH);
+	_(TDT);
+	_(RAL);
+	_(RAH);
+	}
+	E("Unsupported register: 0x%x", r);
+	k_oops();
+	return NULL;
+}
+
+static enum ethernet_hw_caps e1000_caps(struct device *dev)
+{
+	return  ETHERNET_LINK_10BASE_T | ETHERNET_LINK_100BASE_T | \
+		ETHERNET_LINK_1000BASE_T;
+}
+
+static size_t e1000_linearize(struct net_pkt *pkt, void *buf, size_t bufsize)
+{
+	size_t len = net_pkt_ll_reserve(pkt) + pkt->frags->len;
+	struct net_buf *nb;
+
+	/* First fragment contains link layer (Ethernet) header */
+	memcpy(buf, net_pkt_ll(pkt), len);
+
+	for (nb = pkt->frags->frags; nb; nb = nb->frags) {
+		memcpy(buf + len, nb->data, nb->len);
+		len += nb->len;
+	}
+
+	return len;
+}
+
+static int e1000_tx(struct e1000_dev *dev, void *data, size_t data_len)
+{
+	dev->tx.addr = POINTER_TO_INT(data);
+	dev->tx.len = data_len;
+	dev->tx.cmd = TDESC_EOP | TDESC_RS;
+
+	W(dev, TDT, 1);
+
+	while (!(dev->tx.sta)) {
+		k_yield();
+	}
+
+	D("tx.sta: 0x%02hx", dev->tx.sta);
+
+	return (dev->tx.sta & TDESC_STA_DD) ? E1000_SUCCESS : -EIO;
+}
+
+static int e1000_send(struct net_if *iface, struct net_pkt *pkt)
+{
+	struct e1000_dev *dev = net_if_get_device(iface)->driver_data;
+
+	size_t len = e1000_linearize(pkt, dev->txb, sizeof(dev->txb));
+
+	int status = e1000_tx(dev, dev->txb, len);
+
+	if (status == E1000_SUCCESS) {
+		net_pkt_unref(pkt);
+	}
+
+	return status;
+}
+
+static struct net_pkt *e1000_rx(struct e1000_dev *dev)
+{
+	struct net_pkt *pkt = NULL;
+	union {
```

Comment:
If you never use dw[1], why these complications with union? Just cast u64_t to u32_t where needed.

Reply:
Using a union is the standard way to do type punning.

Reply:
> Using a union is the standard way to do type punning.

Yes, if you need to go e.g. from float to its bits, as int's. Here again, just a narrowing cast from u64_t to u32_t would be enough. Up to you whether you want to change it, mostly a note.

Reply:
> Up to you whether you want to change it, mostly a note.

I had another look at it, and dropped that union.